commons/src/main/java/org/infinispan/commons/dataconversion/BinaryEncoder.java

```diff
@@ -0,0 +1,93 @@
+package org.infinispan.commons.dataconversion;
+
+import java.io.IOException;
+
+import org.infinispan.commons.CacheException;
+import org.infinispan.commons.marshall.StreamingMarshaller;
+import org.infinispan.commons.marshall.WrappedByteArray;
+
+/**
+ * Encoder for StoreType.BINARY. For String and primitives, store unencoded. For other objects, store them marshalled.
+ * @since 9.1
+ */
+public class BinaryEncoder implements Encoder {
+
+   private final StreamingMarshaller marshaller;
+
+   public BinaryEncoder(StreamingMarshaller marshaller) {
+      this.marshaller = marshaller;
+   }
+
+   private Object toStorage(Object content) {
+      try {
+         return skipEncoding(content) ? content : marshall(content);
+      } catch (IOException | InterruptedException e) {
+         throw new CacheException(e);
+      }
+   }
+
+
+   private Object fromStorage(Object stored) {
+      try {
+         if (stored == null || isTypeExcluded(stored.getClass())) {
+            return stored;
+         }
+         return stored instanceof byte[] ? marshaller.objectFromByteBuffer((byte[]) stored) : stored;
+      } catch (IOException | ClassNotFoundException e) {
+         throw new CacheException(e);
+      }
+   }
+
+   @Override
+   public Object keyToStorage(Object content) {
+      return toStorage(content);
+   }
+
+   @Override
+   public Object valueToStorage(Object content) {
+      return toStorage(content);
+   }
+
+   @Override
+   public Object keyFromStorage(Object content) {
+      return fromStorage(content);
+   }
+
+   @Override
+   public Object valueFromStorage(Object content) {
+      return fromStorage(content);
+   }
+
+   @Override
+   public boolean isStorageFormatFilterable() {
+      return false;
+   }
+
+   @Override
+   public Object extractIndexable(Object content) {
+      return fromStorage(content);
+   }
+
+
+   public static boolean isTypeExcluded(Class<?> type) {
+      return type.equals(String.class) || type.isPrimitive() ||
+            type.equals(Boolean.class) || type.equals(Character.class) ||
+            type.equals(Byte.class) || type.equals(Short.class) || type.equals(Integer.class) ||
+            type.equals(Long.class) || type.equals(Float.class) || type.equals(Double.class) ||
+            // We cannot exclude array as we can't tell the difference between byte[] and WrappedByteArray
+            type.equals(WrappedByteArray.class);
```

Comment:
I don't think we need this anymore, due to the `Wrapper` being separated from the `Encoder` but you would have to try that out :)